framework/src/main/java/org/checkerframework/framework/source/CompositeChecker.java

```diff
@@ -0,0 +1,756 @@
+   * @param paramTypes the types of the constructor's parameters
+   * @param args the arguments on which to invoke the constructor
+   * @return the result of the constructor invocation on {@code args}, or null if the class does not
+   *     exist
+   */
+  @SuppressWarnings({"unchecked", "TypeParameterUnusedInFormals"}) // Intentional abuse
+  public static <T> @Nullable T invokeConstructorFor(
+      @ClassGetName String name, Class<?>[] paramTypes, Object[] args) {
+
+    // Load the class.
+    Class<T> cls;
+    try {
+      cls = (Class<T>) Class.forName(name);
+    } catch (Exception e) {
+      // no class is found, simply return null
+      return null;
+    }
+
+    assert cls != null : "reflectively loading " + name + " failed";
+
+    // Invoke the constructor.
+    try {
+      Constructor<T> ctor = cls.getConstructor(paramTypes);
+      return ctor.newInstance(args);
+    } catch (Throwable t) {
+      if (t instanceof InvocationTargetException) {
+        Throwable err = t.getCause();
+        if (err instanceof UserError || err instanceof TypeSystemError) {
+          // Don't add more information about the constructor invocation.
+          throw (RuntimeException) err;
+        }
+      } else if (t instanceof NoSuchMethodException) {
+        // Note: it's possible that NoSuchMethodException was caused by
+        // `ctor.newInstance(args)`, if the constructor itself uses reflection.
+        // But this case is unlikely.
+        throw new TypeSystemError(
+            "Could not find constructor %s(%s)", name, StringsPlume.join(", ", paramTypes));
+      }
+
+      Throwable cause;
+      String causeMessage;
+      if (t instanceof InvocationTargetException) {
+        cause = t.getCause();
+        if (cause == null || cause.getMessage() == null) {
+          causeMessage = t.getMessage();
+        } else if (t.getMessage() == null) {
+          causeMessage = cause.getMessage();
+        } else {
+          causeMessage = t.getMessage() + ": " + cause.getMessage();
+        }
+      } else {
+        cause = t;
+        causeMessage = (cause == null) ? "null" : cause.getMessage();
+      }
+      throw new BugInCF(
+          cause,
+          "Error when invoking constructor %s(%s) on args %s; cause: %s",
+          name,
+          StringsPlume.join(", ", paramTypes),
+          Arrays.toString(args),
+          causeMessage);
+    }
+  }
+
+  /**
+   * Returns the requested subchecker. A checker of a given class can only be run once, so this
+   * returns the only such checker, or null if none was found. The caller must know the exact
+   * checker class to request.
+   *
+   * @param <T> the class of the subchecker to return
+   * @param checkerClass the class of the subchecker to return
+   * @return the requested subchecker or null if not found
+   */
+  @SuppressWarnings("unchecked")
+  public <T extends BaseTypeChecker> @Nullable T getSubchecker(Class<T> checkerClass) {
```

Comment:
Why do subcheckers have to extend `BaseTypeChecker`? Why can't they extend `CompositeChecker` instead?

Reply:
@kelloggm I've addressed your comments, let me know what you think!

This specific question I think I answered above already:



"I wanted to change as little as possible for the BaseTypeChecker. Composite checker is subclassed directly exclusively by the refactored RLC. However, many classes with subcheckers subclass BaseTypeChecker directly. They want their subcheckers to be BaseTypeCheckers. Would the subcheckers be CompositeCheckers, they would have to cast them to access their typefactories and analyses.

I know, it's a bit unfortunate, but I didn't find a better solution."

Reply:
And thank you for the feedback!